compiler/src/dotty/tools/dotc/transform/TailRec.scala

```diff
@@ -100,62 +96,58 @@ class TailRec extends MiniPhase {
         cpy.DefDef(dd)(rhs = {
           val defIsTopLevel = sym.owner.isClass
           val origMeth = sym
-          val label = mkLabel(sym)
           val owner = ctx.owner.enclosingClass.asClass
 
-          var rewrote = false
-
           // Note: this can be split in two separate transforms(in different groups),
           // than first one will collect info about which transformations and rewritings should be applied
           // and second one will actually apply,
           // now this speculatively transforms tree and throws away result in many cases
-          val rhsSemiTransformed = {
-            val transformer = new TailRecElimination(origMeth, owner, mandatory, label)
-            val rhs = transformer.transform(dd.rhs)
-            rewrote = transformer.rewrote
-            rhs
-          }
-
-          if (rewrote) {
-            if (tree.symbol.owner.isClass) {
-              val classSym = tree.symbol.owner.asClass
-
-              val labelDef = DefDef(label, vrefss => {
-                assert(vrefss.size == 1, vrefss)
-                val vrefs = vrefss.head
-                val thisRef = vrefs.head
-                val origMeth = tree.symbol
-                val origVParams = vparams.map(_.symbol)
+          val transformer = new TailRecElimination(origMeth, owner, vparams.map(_.symbol), mandatory)
+          val rhsSemiTransformed = transformer.transform(dd.rhs)
+
+          if (transformer.rewrote) {
+            val varForRewrittenThis = transformer.varForRewrittenThis
+            val rewrittenParamSyms = transformer.rewrittenParamSyms
+            val varsForRewrittenParamSyms = transformer.varsForRewrittenParamSyms
+
+            val initialValDefs = {
```

Comment:
Good question. Depends on what the coding style in Dotty is. There are `var` definitions, but they are still `tpd.ValDef`s because there is no such thing as a `VarDef`. That's why I called them `initialValDefs`. Is it customary to refer to `tpd.ValDef`s that represents `var`s as `VarDef`s?